documentation/proposals/Proposal - Multi-Backend Input.md

```diff
@@ -11,203 +11,683 @@ Proposal API for backend-agnostic, refactored Input via keyboards, mice, and con
+    public event Action<AxisEvent<GamepadState, Vector2>>? ThumbstickMove;
+    public event Action<AxisEvent<GamepadState, float>>? TriggerMove;
+}
+
+public partial class Joysticks : IReadOnlyList<JoystickState>
+{
+    public ref readonly JoystickState this[int index] { get; }
+    JoystickState IReadOnlyList<JoystickState>.this[int index] { get; }
+    
+    public <implementation defined> GetEnumerator();
+    IEnumerator<JoystickState> IEnumerable<JoystickState>.GetEnumerator();
+
+    public event Action<ButtonEvent<JoystickState, JoystickButton>>? ButtonDown;
+    public event Action<ButtonEvent<JoystickState, JoystickButton>>? ButtonUp;
+    public event Action<AxisEvent<JoystickState, float>>? AxisMove;
+    public event Action<AxisEvent<JoystickState, Vector2>>? HatMove;
+}
+```
+
+Unlike 1.0 and 2.0, this proposal uses `readonly record struct`s as their only argument for the event action. This allows us to provide more information to the event handlers without breaking in the future. These types are farily simple:
+
+```cs
+public readonly record struct ButtonEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button);
+public readonly record struct ClickEvent<TDevice, TButton>(in TDevice DeviceState, TButton Button, Vector2 Position);
+public readonly record struct AxisEvent<TDevice, TAxis>(in TDevice DeviceState, int Index, TAxis OldValue, TAxis NewValue);
+public readonly record struct TextEvent<TDevice>(in TDevice DeviceState, string? OldText, string? NewText);
+```
+
+This is the part of this proposal that incorporates the ideas in Enhanced Input Events, and is why this proposal supersedes that one.
+
+One final point to note is that throughout the rest of the proposal the following type will be used:
+
+```cs
+public struct InputReadOnlyList<T> : IReadOnlyList<T>
+{
+    public InputReadOnlyList(IReadOnlyList<T> other);
+}
+```
+
+The Silk.NET team wishes to reserve the right to add more constructors to this type as it sees fit.
+
+This exists so that, should the Silk.NET choose to, we can optimize the lookup of elements while ensuring things like indexers are inlined and don't result in a virtual call if our implementation allows us to do so.
+
+**INFORMATIVE TEXT:** For example, for joystick and mouse buttons we could use a fixed-sized bit buffer where each bit represents an individual button: 1 for pressed, 0 for unpressed. But for something like keyboard input where there are a large amount of keys, we can't do that and will likely use `Memory<T>` instead.
+
+# Mouse Input
+
+As discussed earlier, the interface will be very simple.
+
+```cs
+public interface IMouse : IInputDevice
+{
+    MouseState State { get; }
+    ICursorConfiguration Cursor { get; }
+    void SetPosition(Vector2 pos);
+}
+```
+
+`State` is the device state as defined earlier.
+
+`Cursor` contains the cursor configuration. This isn't actually state that the end user can change, and has been made an interface rather than a state struct accordingly.
```

Comment:
🤔 is that true? Can the user not change the cursor configuration? I'm pretty sure they could
This is a bit ridiculous but why can't they use this menu for example:
![image](https://user-images.githubusercontent.com/22711887/151716308-b550b6b8-079c-4282-8cd2-aeba84ca5f58.png)

Reply:
Those aren't any of the configuration options we support or our underlying APIs support. Those are actually exclusively user options, too, and can't be easily queried to my knowledge.